### ISSUE TITLE
Inconsistent behaviour when updating a typed dict with a "subset" typeddict

### ISSUE THREAD

**Issue by Alexandre-SCHOEPP** [python/mypy#19365]
## Context

I am handling configuration from various sources, including kwargs. To do this, I use the `TypeDict`s, and I got this unexpected error:

```python
# minimum working example
from typing_extensions import TypedDict, NotRequired


class Config(TypedDict):
    a: str
    b: float


class UserOverride(TypedDict):
    a: NotRequired[str]


def get_default_config(user_override: UserOverride) -> Config:
    conf: Config = {
        "a": "hello world",
        "b": 1.
    }
    conf.update(user_override)
    conf.update(**user_override)  # Extra argument "a" for update of TypedDict [misc]
    return conf
```

The first `update` call does not raise any error, but the second, which is equivalent according to all python rules I known, does not function.

If mypy is enforcing a standard way of writing code (which may be the case), then the error message is horribly unhelpful, pointing me to go through my complex library to seek the various `TypedDict`s  definitions and try and find an incompatibility that does not exist.

I also disagree with the assessment that one signature is better than the other. Pylance errors on the first `update` call, telling me that it does not match an overload (I am probably going to open an issue with them about it too).

## System info

* OS: Windows 11
* Mypy: mypy 1.16.0 (compiled: yes)
* Python: 3.12.3

**Comment by sterliakov** [comment]
Note that `.update(user_override)` is also rejected under `--strict`:

https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=cebb6c33ceba44aa58e652e244247ff3

**Comment by Alexandre-SCHOEPP** [comment]
> Note that .update(user_override) is also rejected under --strict:
>
> https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=cebb6c33ceba44aa58e652e244247ff3

I don't understand this choice. Why shouldn't I update a subset of keys of my typeddict ?

**Comment by brianschubert** [comment]
Both calls are unsafe, since `user_override` could have arbitrary extra items that could conflict with `Config`. For example, `{"b": "bad"}` is a valid `UserOverride`. [PEP 728](https://peps.python.org/pep-0728/) should make it possible to type this safely (if accepted).

**Comment by Alexandre-SCHOEPP** [comment]
> Both calls are unsafe, since user_override could have arbitrary extra items that could conflict with Config. [PEP 728](https://peps.python.org/pep-0728/) should make it possible to type this safely.

Can mypy not establish at runtime that the all keys of `UserOverride`  (and by extent, `user_override`) are in `Config` ?
Wouldn't the only unsafe aspect be when I define both dicts ?

I understand that it would be unsafe, were `UserOverride` not to be a Typeddict

> For example, {"b": "bad"} is a valid UserOverride. 

[Not according to mypy](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=f005a865fa39d4bfc84d240789e5065c)

**Comment by brianschubert** [comment]
> [Not according to mypy](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=f005a865fa39d4bfc84d240789e5065c)

Construction is a special-case outlined in the typing spec. I believe the idea was to help catch issues like typos in key names, though it has lead to some confusion due to its incongruence with the usual assignability rules. You can refer to the [TypedDict Assignability](https://typing.python.org/en/latest/spec/typeddict.html#assignability) section of the typing spec for the how TypedDict compatibility works in general.

**Comment by Alexandre-SCHOEPP** [comment]
Ok so it's not Mypy that is inconsistent, but rather Python 😫

**Comment by sterliakov** [comment]
> Ok so it's not Mypy that is inconsistent, but rather Python 😫

Well, not that you should need typed dicts often. Such structural subtyping has problems that simply can't be resolved without knowledge about keys that aren't mentioned.

(I'll be mostly retelling the spec below)

The spec for TypedDict is at least sound. Consider the following chain of transformations:

```python
class Full(TypedDict):
    a: str
    b: float

class Part(TypedDict):
    a: NotRequired[str]
    
class WrongPart(TypedDict):
    a: NotRequired[str]
    b: NotRequired[list[object]]


base: Full = {'a': '', 'b': 0}

part: WrongPart = {'b': []}
part_up: Part = part
base.update(part_up)
```

If both of them are allowed, you just ended up having a list under key `b` of type `float`. Not cool. So at least one of two last lines must be a type error. Naturally you want the upcast to succeed ("I only care about key `a`", it would be impossible to pass - even readonly - wider typeddicts to functions accepting narrow argument). So the lat line must be a type error, right?

Okay, there's a third option: say "who cares about type safety" (khm-khm TypeScript that [just allows my example verbatim](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgGIFcA2mDeAoZQ5OALmQGcwpQBzAbgKICMyR0BbJ6BgXz1EixEKAApwoYfEWIB+MpWoh6ePtP7ho8JMgDqUAPZKxEqUThyKVWg2lML6EAGsQ+gO4gA2gF1eePAkNKZCY4cggyDGxkAF5kHAByOHiyePiAGmR4pmTkAAY+AJAggAdxMDI9QxpjMBi4rJzvHgZCkrKAfXRishq60okWwNr2aBoIABMIrEw6nAA6BZCwjIW5-rBO4ua8AHod5AAVAAs4J3IMl1dkMBPa8Sg4AE8KCAh2cmv9ZBOANxQf07AchHCYyfxDZDtVgQP5QOojKBjcZzJh0Qh7Q6PYooeJsTjQeLIIHIFx3cjkYA0EBwJiYFBgL5gLE4kAwglzPBAA)). It's definitely not something I expect from a type system of a mature and convenient language - such as python.

**Comment by Alexandre-SCHOEPP** [comment]
**Note**: I understand that it isn't an issue with mypy, and will close the issue with this comment

> If both of them are allowed, you just ended up having a list under key `b` of type `float`. Not cool.

I personally feel like the transformation that fails the type safety here is the addition of the new key to `Part`

In your example, for example, 

```python
part: WrongPart = {'b': []}
part_up: Part = part  # The spec allows this, but I don't think it should.
base.update(part_up)  # ergo this should not occur, and the static checker should tell you that WrongPart cannot update Full
```

I have never had a usecase where I wanted to define a typeddict that could accept keys outside of its definition. I feel like allowing that to occur is only going to lead to spelling mistakes becoming silent errors.

>Well, not that you should need typed dicts often.

Given that in the type-hinting system, kwargs are defined as Typeddicts, I actually need it as soon as my function has to handle kwargs, which occurs whenever I have overloads, or user-configurations that overrides local defaults, which overrides global defaults (a common practice in my project).

I actually define and combine 17 different typeddicts in my 35k lines project, and this low of an amount is only because I usually spell all kwargs out.

## Conclusion

All in all, I fundamentally disagree with the logic of "You can't _define_ with extra keys, but you can _add them later_" of the spec.
Type-strictness as an opt-in is already the default in python. If you don't want it, why use such an advanced feature ?
